### PR TITLE
review feat: get imports from source code at first in auto-import mode

### LINKED ISSUE #1453: autoimport: start with the original import list?
in autoimport mode, we may be able

* start with the initial import list from the original source file 
* remove the imports not used anymore due to transformation
* add the new imports due to transformation

### PR DESCRIPTION
**Pull request by surli**
Use the imports from source code in auto-import mode should fix #1453

### CODE REVIEW

**Comment by tdurieux** [comment]
Please don't use reflection to replace the star import.

**Comment by surli** [comment]
> Please don't use reflection to replace the star import.

Fixed.

File: src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Created by urli on 08/08/2017.
+ */
+public class JDTImportBuilder {
```
Comment (monperrus) [inline src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java]: package-protected?

File: src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	private ICompilationUnit sourceUnit;
+	private Factory factory;
+	private Set<CtReference> imports;
+
+	JDTImportBuilder(CompilationUnitDeclaration declarationUnit,  Factory factory) {
+		this.declarationUnit = declarationUnit;
+		this.factory = factory;
+		this.sourceUnit = declarationUnit.compilationResult.compilationUnit;
+		this.filePath = CharOperation.charToString(sourceUnit.getFileName());
+		this.spoonUnit = factory.CompilationUnit().create(filePath);
```
Comment (monperrus) [inline src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java]: document that's it's not a create but a get from cache after traverse

File: src/main/java/spoon/reflect/cu/CompilationUnit.java [hunk 3]
```diff
@@ -107,4 +109,8 @@
 	 */
 	int getTabCount(int index);
 
+	Set<CtReference> getImports();
```
Comment (monperrus) [inline src/main/java/spoon/reflect/cu/CompilationUnit.java]: API doc?

File: src/main/java/spoon/reflect/cu/CompilationUnit.java [hunk 4]
```diff
@@ -107,4 +109,8 @@
...
+
+	void setImports(Set<CtReference> imports);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/cu/CompilationUnit.java]: API doc?

File: src/main/java/spoon/reflect/cu/CompilationUnit.java (hunk 3 above)
Comment (monperrus) [inline src/main/java/spoon/reflect/cu/CompilationUnit.java]: replace with Collection?

File: src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+		if (klass == null) {
+			klass = this.factory.Interface().get(className);
+
+			if (klass == null) {
+				try {
+					Class zeClass = this.getClass().getClassLoader().loadClass(className);
+
+					if (zeClass.isInterface()) {
+						klass = this.factory.Interface().get(zeClass);
```
Comment (surli) [inline src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java]: Change to get from Type()

File: src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+		for (ImportReference importRef : declarationUnit.imports) {
+			String importName = importRef.toString();
+			if (!importRef.isStatic()) {
+				// Starred import are only managed by importing types of the model for now
+				// as it can cost a lot to retrieve all classes of a package by reflection
+				if (importName.endsWith("*")) {
+					int lastDot = importName.lastIndexOf(".");
+					String packageName = importName.substring(0, lastDot);
+
+					CtPackage ctPackage = this.factory.Package().get(packageName);
```
Comment (monperrus) [inline src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java]: check that this does not use reflection too with shadow packages. if not document

File: src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+				if (klass != null) {
+
+					// for now starred import are treated by importing
+					// all static fields and methods
+					// or all fields and methods if it concerns an interface
+					if (methodOrFieldName.equals("*")) {
+						Collection<CtFieldReference<?>> fields = klass.getAllFields();
+						Set<CtMethod> methods = klass.getAllMethods();
+
+						for (CtFieldReference fieldReference : fields) {
```
Comment (monperrus) [inline src/main/java/spoon/support/compiler/jdt/JDTImportBuilder.java]: check if public?

**Pull request merged**